### ISSUE TITLE
More properties for formal verification

### ISSUE THREAD

**Issue by daejunpark** [morpho-org/morpho-blue#256]
EDIT by @QGarchery. Todo list:
- [x] A1, stay healthy  (in progress, some timeouts left after #344)
- [x] B, roundtrip properties (in progress, first steps in #345)
- [x] D1, interest accrual increases shares value (in progress in #353)

A2 is omitted because it does not seem to hold. Path independence property are postponed to future work, and they also require accepting a small delta. D2 is already mentioned in #96, D3 is being considered as an improvement on D2.

END OF EDIT

In addition to the nice invariants in #96, these properties can be considered for formal verification:

## Properties about healthiness

**Property A.1**
No Blue transaction, including liquidation calls, can cause a healthy position to become unhealthy.  (In other words, only price changes occurring outside of Blue can result in a previously healthy position becoming unhealthy.)

For example, this property can be tested as follows:
```
vm.assume(_isHealthy(market, id, user));

address(blue).call(data);

assert(_isHealthy(market, id, user));
```

**Property A.2** _[Updated: 8/18]_
"Partial" liquidation cannot change an unhealthy position into a healthy one, if the LTV is greater than `1 / liquidation_incentive`.

For example,
```
vm.assume(ultimatelyInsolvent(market, id, borrower));

blue.liquidate(market, borrower, seized, data);

borrowAmount = borrowShares[id][borrower];
isHealthy = _isHealthy(market, id, borrower);

// borrowAmount > 0  implies  !isHealthy
assert(borrowAmount == 0 || !isHealthy);
```

where:
```
function ultimatelyInsolvent(Market memory market, Id id, address user) internal view returns {
    uint256 borrowed = borrowShares[id][user].toAssetsUp(totalBorrow[id], totalBorrowShares[id]);            

    uint256 collateralPrice = IOracle(market.oracle).price();
    uint256 collateralValue = collateral[id][user].mulDivDown(collateralPrice, ORACLE_PRICE_SCALE);

    uint256 incentive = UtilsLib.min(
        MAX_LIQUIDATION_INCENTIVE_FACTOR, WAD.wDivDown(WAD - LIQUIDATION_CURSOR.wMulDown(WAD - market.lltv))
    );

    return borrowed.wMulDown(incentive) > collateralValue;
}
```

### Roundtrip properties

Roundtrip properties between supply() and withdraw():
- **Property B.1.1**: There should be no profit from supply() immediately followed by withdraw().
- **Property B.1.2**: There should be no profit from withdraw() immediately followed by supply().

Similarly, roundtrip properties between borrow() and repay():
- **Property B.2.1**: There should be no profit from borrow() immediately followed by repay().
- **Property B.2.2**: There should be no profit from repay() immediately followed by borrow().

For example, Property B.1.1 can be tested as follows:
```
Id id = market.id();

uint oldBalance = IERC20(market.borrowableAsset).balanceOf(x);
uint oldShares = blue.supplyShares(id, x);

blue.supply(market, amount, x, data);
uint shares = blue.supplyShares(id, x) - oldShares;
blue.withdraw(market, shares, x, data);

uint newBalance = IERC20(market.borrowableAsset).balanceOf(x);
assert(newBalance <= oldBalance);
```

### Path independence properties

**Property C.1**
Path independence for `_accrueInterests()`:
- Multiple interest accruals do not generate more interest than a single accrual for the same time period.

For example,
```
current = lastUpdate[id];
oldSupply = totalSupply[id];
vm.warp(current + x);
_accrueInterests(market, id);
vm.warp(current + x + y);
_accrueInterests(market, id);
delta1 = totalSupply[id] - oldSupply;

current = lastUpdate[id];
oldSupply = totalSupply[id];
vm.warp(current + x + y);
_accrueInterests(market, id);
delta2 = totalSupply[id] - oldSupply;

assert(delta1 <= delta2);
```

**Property C.2**
Path independence for `liquidate()`
- There is no benefit to making multiple liquidate() calls compared to a single liquidate() call for the same total seized amount.

For example,
```
oldBalance = IERC20(market.borrowableAsset).balanceOf(caller);
vm.prank(caller); blue.liquidate(market, borrower, x, data);
vm.prank(caller); blue.liquidate(market, borrower, y, data);
repaid1 = oldBalance - IERC20(market.borrowableAsset).balanceOf(caller);

oldBalance = IERC20(market.borrowableAsset).balanceOf(caller);
vm.prank(caller); blue.liquidate(market, borrower, x + y, data);
repaid2 = oldBalance - IERC20(market.borrowableAsset).balanceOf(caller);

assert(repaid1 >= repaid2);
```

### Properties about share prices

**Property D.1**
The price of both supply and borrow shares do not decrease after interest accruals.

For example,
```
oldSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
oldBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

_accrueInterests(market, id);

newSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
newBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

assert(oldSupplySharePrice <= newSupplySharePrice);
assert(oldBorrowSharePrice <= newBorrowSharePrice);
```

**Property D.2**
More generally, the price of both supply and borrow shares do not decrease after any non-liquidation Blue transactions.

For example,
```
oldSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
oldBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

vm.assume(selector != blue.liquidate.selector);
address(blue).call(abi.encodePacked(selector, data));

newSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
newBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

assert(oldSupplySharePrice <= newSupplySharePrice);
assert(oldBorrowSharePrice <= newBorrowSharePrice);
```

**Property D.3**
The share prices do not decrease after liquidation, as long as it wasn't a result of bad debt.

For example,
```
oldSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
oldBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

blue.liquidate(market, borrower, seized, data);
vm.assume(collateral[id][borrower] > 0); // no bad debt

newSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
newBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

assert(oldSupplySharePrice <= newSupplySharePrice);
assert(oldBorrowSharePrice <= newBorrowSharePrice);
```

**Property D.4** _[Added 8/18]_
The share prices do not decrease after liquidation, if the LTV is less than or equal to `1 / liquidation_incentive`.

For example,
```
oldSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
oldBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

vm.assume(ultimatelyInsolvent(market, id, borrower) == false);  // ultimately solvent, implying no bad debt

blue.liquidate(market, borrower, seized, data);

newSupplySharePrice = toAssetsDown(1e18, totalSupply[id], totalSupplyShares[id]);
newBorrowSharePrice = toAssetsDown(1e18, totalBorrow[id], totalBorrowShares[id]);

assert(oldSupplySharePrice <= newSupplySharePrice);
assert(oldBorrowSharePrice <= newBorrowSharePrice);
```
where `ultimatelyInsolvent()` is defined in Property A.2.
